java/core/src/java/org/apache/orc/impl/writer/TimestampInstantTreeWriter.java

```diff
@@ -0,0 +1,160 @@
+import org.apache.hadoop.hive.ql.exec.vector.TimestampColumnVector;
+import org.apache.hadoop.hive.ql.util.JavaDataModel;
+import org.apache.orc.OrcProto;
+import org.apache.orc.TypeDescription;
+import org.apache.orc.impl.IntegerWriter;
+import org.apache.orc.impl.PositionRecorder;
+
+import java.io.IOException;
+import java.text.DateFormat;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.TimeZone;
+
+import static org.apache.orc.impl.writer.TimestampTreeWriter.MILLIS_PER_SECOND;
+
+public class TimestampInstantTreeWriter extends TreeWriterBase {
+  private final IntegerWriter seconds;
+  private final IntegerWriter nanos;
+  private final boolean isDirectV2;
+  private final long baseEpochSecsUTC;
+
+  public TimestampInstantTreeWriter(int columnId,
+                                    TypeDescription schema,
+                                    WriterContext writer,
+                                    boolean nullable) throws IOException {
+    super(columnId, schema, writer, nullable);
+    this.isDirectV2 = isNewWriteFormat(writer);
+    this.seconds = createIntegerWriter(writer.createStream(id,
+        OrcProto.Stream.Kind.DATA), true, isDirectV2, writer);
+    this.nanos = createIntegerWriter(writer.createStream(id,
+        OrcProto.Stream.Kind.SECONDARY), false, isDirectV2, writer);
+    if (rowIndexPosition != null) {
+      recordPosition(rowIndexPosition);
+    }
+    DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+    dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+    try {
+      this.baseEpochSecsUTC = dateFormat
+          .parse(TimestampTreeWriter.BASE_TIMESTAMP_STRING).getTime() /
+          MILLIS_PER_SECOND;
+    } catch (ParseException e) {
+      throw new IOException("Unable to create base timestamp instant tree writer", e);
+    }
+  }
+
+  @Override
+  OrcProto.ColumnEncoding.Builder getEncoding() {
+    OrcProto.ColumnEncoding.Builder result = super.getEncoding();
+    if (isDirectV2) {
+      result.setKind(OrcProto.ColumnEncoding.Kind.DIRECT_V2);
+    } else {
+      result.setKind(OrcProto.ColumnEncoding.Kind.DIRECT);
+    }
+    return result;
+  }
+
+  @Override
+  public void writeBatch(ColumnVector vector, int offset,
+                         int length) throws IOException {
+    super.writeBatch(vector, offset, length);
+    TimestampColumnVector vec = (TimestampColumnVector) vector;
+    if (vector.isRepeating) {
+      if (vector.noNulls || !vector.isNull[0]) {
+        // ignore the bottom three digits from the vec.time field
+        final long secs = vec.time[0] / MILLIS_PER_SECOND;
+        final int newNanos = vec.nanos[0];
+        // set the millis based on the top three digits of the nanos
+        long millis = secs * MILLIS_PER_SECOND + newNanos / 1_000_000;
+        if (millis < 0 && newNanos > 999_999) {
+          millis -= MILLIS_PER_SECOND;
+        }
+        indexStatistics.updateTimestamp(millis);
```

Comment:
there seems to be some code duplication here. except for millis and seconds rest all looks the same between both timestamp writers. should the common code be factored into a common method? in case of any fixes will be easier to fix in one place. similarly for the readers.